### PR TITLE
migrate `Float` shrinker to the ir

### PR DESCRIPTION
**Pull request by tybug**
I had originally tried writing a `shrinker_v2.py` from scratch...and quickly abandoned it as infeasible. I think my current plan is to convert shrinking passes one by one (where possible, which may or may not be all of them). `minimize_floats` is one of the more directly applicable ones. 

Definitely open to feedback on the details of the migration approach here. 

There is overlap in the `IRTree` definition with #3806, though it's not exactly the same. Either is fine to merge before the other; I'll handle conflict resolution whichever way.

### CODE REVIEW

File: hypothesis-python/src/hypothesis/internal/conjecture/data.py
Comment (tybug) [inline hypothesis-python/src/hypothesis/internal/conjecture/data.py:921]: Replacement for `Example.trivial` / `Blocks.trivial`. I think I'll actually remove this before merge, as I don't use it for `Float` (should I? we didn't previously), but I expect to use this general approach in future shrink pass conversions.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py
Comment (tybug) [inline hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py:385]: this triples our shrink cost...we could feasibly get it down to double, since the second pass is for maintaining the correct `forced` values (which affects `block.trivial`). It's just annoying and messy to thread a `fake_forced` parameter through everywhere.

At some point, more than half of our passes will be on the ir, at which point it's more efficient to convert from bits to ir when necessary, and maintain the tree as the source of truth. So in practice, this caps out at 1.75x slowdown or so at the inflection point.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinking/floats.py
Comment (tybug) [inline hypothesis-python/src/hypothesis/internal/conjecture/shrinking/floats.py:38]: a nice intermediate bonus from the ir: we know the bounds, so we can throw away invalid tries.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py:385]: I think that's worth it; 3x is a very heavy cost to pay.

Could we also drop the cost of this first stage by only running the IR logic if there's a float in the tree?  That way there's no slowdown (well, small constant) for unaffected tests.

Comment (tybug) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py:385]: I think that's already the case: `consider_new_tree` is only called when `Float.shrink` runs, which is gated behind a `leaf = chooser.choose(self.leaves, lambda leaf: leaf.ir_type == "float")`.

Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py:385]: Oh, nice, if that's worst-case instead of average it's not such a big deal; hopefully also offset by high quality/specificity shrinking.

**Comment by tybug** [comment]
Seeing slow shrinks locally. I think I'll need to add ir_tree support to `cached_test_function`, alongside buffers. We'll maintain both and then drop  buffers once everything is on the ir.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py
Comment (tybug) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py:385]: Well, good news is I was being foolish / overzealous and the extra third call isn't necessary 😄. We're dealing purely with the buffer, so who cares about the forced status of the blocks.

File: hypothesis-python/src/hypothesis/internal/conjecture/data.py
Comment (tybug) [inline hypothesis-python/src/hypothesis/internal/conjecture/data.py:1937]: shrinking may make modifications that misalign the ir tree. When working with buffers, we still run these all the way to a result, but my guess is they are very unlikely to be useful shrinks.

I don't know if there's anything better we can do for ir trees than giving up, beyond realigning tricks that are research papers in their own right. One idea is passing `shrinker.engine.random` here and continuing randomly, which might work, but I don't know if the same ir tree prefix running to different buffers will blow up. 

By bailing out early we do less work in the test function, which might mean we actually spend more time proportionally on useful shrinks by overrunning. So it might be a wash at the end of the day. But I don't have a good intuition.

**Comment by tybug** [comment]
Good and bad news, mostly good:

* `cached_test_function` actually already caches our `ConjectureData.for_ir_tree` datas, because running an `ir_tree=` data is no different from a `prefix=` data from the outside; both run to a buffer. No need to add a separate cache for ir trees. yay!
* ...as a consequence, this forces — no pun intended — our hand in implementing `fake_forced`. The `ConjectureResult` returned by an `ir_tree=` data needs to be completely correct, because it will be cached as the result for a buffer. `fake_forced` is messy, hard to verify correct, and will probably need to stick around even after the ir migration is complete. oh well, can't have everything. 
* pending ci, float shrinking basically works! There's a single failure in `test_float_shrink_can_run_when_canonicalisation_does_not_work`, which isn't a "float shrinking is broken" failure, but "you are calling the float shrinker on 1000.0 where you didn't used to". related review comment below.

I've pushed all my work here so we can see how everything fits together in a single place, but I'm going to bundle a few smaller changes in a split pr shortly. As is typical, this pr got very large very quickly 😄

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py
Comment (tybug) [inline hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py:1220]: Right now, we call `Float.shrink` on any float node (+ some short circuiting behavior in `Float`). The previous behavior skipped `Float.shrink` on floats that happened to also be integers:

```python
        buf = self.shrink_target.buffer
        b = buf[u:v]
        f = lex_to_float(int_from_bytes(b))
        b2 = int_to_bytes(float_to_lex(f), 8)
        if b == b2 or self.consider_new_buffer(buf[:u] + b2 + buf[v:]):
```

I think the `if b == b2` condition there is equivalent to checking `if is_simple(f)`. i.e., don't try to float-shrink float nodes if they are already integers. The comments in `lexical.py` in [this commit](https://github.com/HypothesisWorks/hypothesis/commit/47f2adc47953619c9a93cfb3dd7578937db3429e#diff-5044b2f8ebb16009a425e4c08a026c87b6611d9986f3b0c16f92b5329c057508L81-L89) give some further insight: 

```python
        # This floating point number can be represented in our simple format.
        # So we try converting it to that (which will give the same float, but
        # a different encoding of it). If that doesn't work then the float
        # value of this doesn't unambiguously give the desired predicate, so
        # this approach isn't useful. If it *does* work, then we're now in a
        # situation where we don't need it, so either way we return here.
        if is_simple(f):
            self.incorporate_float(f)
            return
```

To be honest, I'm not sold that we don't want to shrink float nodes as floats even if they happen to currently be integers. We mostly defer to the `Integer` shrinker anyway, and it also seems to make logical sense to keep all float-shrink-logic in the same place instead of jumping around. But I'm probably missing a good historical reason for it to be done this way. It's easy to put this condition back either way.

Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py:1220]: I think the intuition here is that we consider integer-valued floats simpler than all non-integer-valued floats, and so once we get to an integer-value we might as well use the machinery we have for shrinking integers.

You could probably convince me that a different ordering is better, but that's how I understand the status quo.

Comment (tybug) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py:1220]: ok, I think that makes sense to me. I'll add this safeguard back. I was thinking about the case where shrinking an integer-valued float is faster if you allow it to try intermediate float values. But `Integer` probably gets you close enough to the minimum and then you can finish it off with a final `Float` pass. 

I'd like to take a detailed look at the shrink ordering at some point. I saw a few weird things like values in (0, 1) being ordered after almost everything else; `assert float_to_lex(sys.float_info.max) < float_to_lex(0.5)`. But for another day! 

I think Hypothesis' shrinker is so well tuned that I shouldn't make changes without profiling or being very confident in them first.

Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py:1220]: Values between zero and one: not integers, and we order kinda by number of decimal places, so.  And the largest finite float is an integer, which means "simpler" than 0.5!  (I did say I'm not that attached to the ordering)

**Comment by tybug** [comment]
Unfortunately ran out of time tonight for the aforementioned split pr, but will get to it tomorrow.

**Comment by tybug** [comment]
Now that the infra is in place, I'm hopeful we can fly through these shrinker migrations. Only had time for this tonight, but more pulls to come!

**Comment by tybug** [comment]
ack, coverage failure - will have to wait for tomorrow.

File: hypothesis-python/src/hypothesis/internal/conjecture/data.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/conjecture/data.py:2076]: I think what we care about is that `node.value` is allowed by the arguments to this function - if the kwargs don't match exactly that seems OK to me?

We could even be somewhat flexible about converting between integer-values int/float (/bool?) and str/bytes (via utf-8).

File: hypothesis-python/src/hypothesis/internal/conjecture/data.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/conjecture/data.py:2078]: Given that we have the same type-based assertions at every call site, I'd be inclined to put them inline here instead.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinking/floats.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/conjecture/shrinking/floats.py:33]: Check against `smallest_nonzero_magnitude` too, for subnormals?  (yes, they're really annoying!)

**Review by Zac-HD — COMMENTED** [review]
Looking good!  Small comments below, but this doesn't _look_ like it'll suddenly grow on us...

File: hypothesis-python/src/hypothesis/internal/conjecture/data.py
Comment (tybug) [reply hypothesis-python/src/hypothesis/internal/conjecture/data.py:2076]: hmm, I guess we could do this. I don't know how useful it would be. I'd love to empirically evaluate misalignments on real properties one day 😄

File: hypothesis-python/src/hypothesis/internal/conjecture/data.py
Comment (tybug) [reply hypothesis-python/src/hypothesis/internal/conjecture/data.py:2078]: referring to this? https://github.com/HypothesisWorks/hypothesis/blob/aa8a3a38df90a191109061ac72d631a79c9f0923/hypothesis-python/src/hypothesis/internal/conjecture/data.py#L1848-L1849
Unfortunately those are for mypy's benefit, or they would be inline here.

Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/conjecture/data.py:2078]: Huh, fair enough.  I think it'd be possible to get around that with generics + overloads, but it'd be worse than status quo.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinking/floats.py
Comment (tybug) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinking/floats.py]: this comment sent me down a rabbit hole of implementing `ir_value_permitted` for all ir types, not just float. I think the up-front effort here will pay off when working on future shrink passes. We get to use `ir_value_permitted` both here and when re-aligning trees.

One of these days I'll consolidate all these ir-related functions into a proper `IRType` class with our five concrete instantiations...

**Comment by tybug** [comment]
3ca282e4a583347a307ce0eb6e1413f376e59fcc is a probably-bug-fix: we weren't considering `smallest_nonzero_magnitude` when counting the number of possible float children. This was probably a ticking time bomb for someone trying to use a small-range float strategy without subnormal support, though I haven't tried writing a strategy that would manifest this myself.

**Review by Zac-HD — APPROVED** [review]
Looks great with some very small comments below, please merge when they're addressed (or not) to your satisfaction!

**Comment by tybug** [comment]
thanks for the comments, they were all helpful!

**Comment by Zac-HD** [comment]
Since we've closed #3086, do you want to open an issue to track all the remaining parts of the migrate-to-IR project?  https://github.com/HypothesisWorks/hypothesis/issues/3086#issuecomment-1794016022 is not particularly detailed and I think we could probably break it down further now that we've come this far.

**Comment by tybug** [comment]
have also been thinking that we need a better issue for the IR project. I'll write something up.

**Pull request merged**
